### PR TITLE
Centralize Join Validity Logic

### PR DESCRIPTION
**Pull request by plypaul**
This PR adds a class to centralize the logic that is used to figure out whether a join between data sources using a specific identifier. To resolve circular dependencies, some of the semantics classes were factored out.

### CODE REVIEW

File: metricflow/model/semantics/valid_join.py
Comment (tlento) [inline metricflow/model/semantics/valid_join.py:19]: Got any good ideas for adding explicit tests on the logic in this class? If not I'll write some test cases using the SCD model as a basis, since that change could be a little bit finicky.

**Review by tlento — APPROVED** [review]
Overall this looks great! Please update the inline assertions to eliminate the index out of bounds error before merge. Everything else is optional.

File: metricflow/protocols/semantics.py
Comment (plypaul) [reply metricflow/protocols/semantics.py]: Weird, these were generated by the IDE. Updated.

File: metricflow/model/semantics/valid_join.py
Comment (plypaul) [reply metricflow/model/semantics/valid_join.py]: Updated.

File: metricflow/model/semantics/valid_join.py
Comment (plypaul) [reply metricflow/model/semantics/valid_join.py:20]: In retrospect, this should have a set. To unblock you though, seems best to merge and then handle later.

File: metricflow/model/semantics/valid_join.py
Comment (tlento) [reply metricflow/model/semantics/valid_join.py]: > Not sure what you meant by implicit in call ordering though?

I was confused, this actually gets called first. So the callsite has to make sure the identifier exists in the data source ahead of time, or else that assertion will trip.

**Pull request merged**
